core/attrlist.hpp

```diff
@@ -0,0 +1,177 @@
+// Copyright 2016 Husky Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "base/serialization.hpp"
+#include "core/objlist_data.hpp"
+#include "core/utils.hpp"
+
+namespace husky {
+
+using base::BinStream;
+
+class BaseAttrList {
```

Comment:
The base class is used to hide the template implementation of the derived class, so that derived classes with different template types can be put in the map container, e.g.,
```std::unordered_map<std::string, BaseAttrList*> attrlist_map;```